### DISCUSSION TITLE
stdin usage question (nodejs)

### DISCUSSION THREAD

**Discussion by ankology** [libvips/libvips#2832] [Q&A] [answered]
First of all thanks for the project @jcupitt!
I'm trying to use libvips command line attached to a child process in node js. Basically i recieve a uploaded file from a website, and pass the buffer to a prompt session in server, executing the command to copy `vips copy stdin .png` and writing to stdin in that process. Unfortunately i come across this error: `VipsForeignLoad: buffer is not in a known format. Exit code: 1`, inspecting the source code [foreign.c#L576](https://github.com/libvips/libvips/blob/a9d64bea54c1fba083a9907c66e19333416a0d7e/libvips/foreign/foreign.c#L576) and the site docs section [streaming](https://www.libvips.org/2019/11/29/True-streaming-for-libvips.html) that says "The magic filename "stdin" opens a stream attached to file descriptor 0" i think that libvips is not finding a valid type for the file (in this case, come from stdin). There exists some way to do that? Maybe a way to tell libvips the type manually with stdin like `vips copy stdin[type=X] .png`? Or i'm miss understanding something? Thanks for the help. Regards!

**Comment by jcupitt** [comment, marked as answer]
Hi @ankology,

It should detect the filetype automatically. I see:

```
$ cat k2.jpg | vips copy stdin .png | cat >x.png
```

And the JPG is converted to a PNG successfully.

What version of libvips are you using? What sort of file are you trying to copy? Has your libvips been built with support for that filetype?

**Reply by ankology** [reply]
Answering: I'm using libvips 8.12.2 in Windows. Just trying to copy a jpg file. Yes it have all the includes. 
So, i stop coding in node to try first in cli, using the copy command with a file path instead of stdin works fine, but trying something like `type .\jpg-test.jpg | vips copy stdin result.png` outputs error: `vips_mapfile: unable to CreateFileMapping\n VipsForeignLoad: buffer is not in a known format`. Maybe some issue related to [this question](https://stackoverflow.com/questions/65014352/pipe-libvips-cli-output-to-stdout-in-windows)? Usage in Windows with stdin from cli is possible? Thanks!

**Reply by ankology** [reply]
By example, using imagemagick `type jpg-test.jpg | magick - result.png` works fine, `-` for the piped stdin.

**Reply by jcupitt** [reply]
I tried in bash on windows and it seems to work:

```
jcupi@DESKTOP-HGI6HBR MINGW64 ~/vips-dev-8.12/bin
$ cat ~/Pictures/k2.jpg | ./vips.exe copy stdin .png | cat > x.png

jcupi@DESKTOP-HGI6HBR MINGW64 ~/vips-dev-8.12/bin
$ mspaint.exe x.png
```

But it fails in powershell, I'm not sure why:

```
PS C:\Users\jcupi\vips-dev-8.12\bin> type ~/Pictures/k2.jpg | .\vips.exe copy stdin x.png
CreateFileMapping failed: vips_mapfile: unable to CreateFileMapping
windows error: %1 is not a valid Win32 application.
```

**Reply by ankology** [reply]
Using Windows cmd.exe not seems to work. In Powershell maybe related to [this answer](https://github.com/mozilla/mozjpeg/issues/269#issuecomment-716720574).

**Reply by ankology** [reply]
Using bash on Windows/Linux my code run successfully with node js spawn child process working with both stdin and stdout. Thank you again!
